ringvax/__init__.py

```diff
@@ -47,52 +47,45 @@ def run(self) -> None:
+            n_infections = len(self.query_people())
+
+            if len(infection_queue) == 0 and not passed_max_generations:
+                self.termination = "extinct"
+                break
+            elif len(infection_queue) == 0 and passed_max_generations:
+                self.termination = "max_generations"
+                break
+            elif n_infections == self.params["max_infections"]:
+                self.termination = "max_infections"
+                break
+            elif n_infections > self.params["max_infections"]:
+                raise RuntimeError("Maximum number of infections exceeded")
+
+            # find the person who is infected next
+            # (the queue is time-sorted, so this is the temporally next infection)
             t_exposed, infector = infection_queue.pop(0)
 
+            # otherwise, instantiate this infection, draw who they in turn infect,
+            # and add the infections they cause to the queue, in time order
             id = self.create_person()
             self.generate_infection(id=id, t_exposed=t_exposed, infector=infector)
 
-            if (
-                len(infection_queue) == 0
-                and len(self.get_person_property(id, "infection_times")) == 0
-            ):
-                termination["criterion"] = "extinction"
-                termination["last_complete_generation"] = max(
-                    self.infections[id]["generation"] for id in self.infections
-                )
-            # add the infections they caused to the end of the queue,
-            # unless we are past the number of generations
-            if (
-                self.get_person_property(id, "generation")
-                < self.params["n_generations"]
-            ):
-                # enqueue the next infections, which will be processed unless we hit
-                # the max. # of infections
-                for t in self.get_person_property(id, "infection_times"):
-                    bisect.insort_right(infection_queue, (t, id), key=lambda x: x[0])
-
-            if len(self.query_people()) >= self.params["max_infections"]:
-                termination["criterion"] = "max_infections"
-                min_in_progress = min(
-                    self.infections[parent]["generation"]
-                    for _, parent in infection_queue
-                )
-                termination["last_complete_generation"] = min_in_progress - 1
-
-        if termination["criterion"] == "":
-            termination["criterion"] = "max_generations"
-            termination["last_complete_generation"] = self.params["n_generations"]
-
-        self.termination = termination
+            # if the infector is in the final generation, do not add their
+            # infectees to the queue
+            generation = self.get_person_property(id, "generation")
+            if generation == self.params["n_generations"]:
+                passed_max_generations = True
+                continue
```

Comment:
This concerns me as well.

I am fully comfortable with what we had earlier: if the infection is in the final generation, don't add its secondaries to the list.

This stops us when we go to process the temporally-first person in the final generation. But that could leave any number of generations incomplete.

Reply:
Cartoon example.
- `max_generations = 2`
- Index case produces infections at t=0.5, and t=2
- Infection at time t=0.5 (generation = 1) produces infection at time 1
- Infection at time 1 (generation = 2) is processed
- termination condition hit

We've got an entire subtree we never started.

Reply:
This speaks more generally to the tight line we're walking between time- and generation-based approaches

Reply:
I don't think this is a problem? `continue` means "go onto the next iteration of the loop", which is different from `break` (exit the loop now).

Like, if we hit the maximum number of infections, we `break`, but here I just `continue`, which means skipping the stuff below it in the loop, which is the part where we add the children of this infection onto the queue.

Maybe it's clearer to write this as `if at_max_gens then set_flag else process_onward_infections`?

Reply:
I made a proposed stopping logic change that I think emphasizes that it's stopping at 0, but I also think `at_max_generations` is clearer than `passed_max_generations`, we can leave the `continue`

Reply:
See note below; I added some more comments and refactored to remove the `continue`